# Algerian Forest Fires Dataset Project EDA Part

## 1. UNDERSTANDING THE GIVEN TASK
#### Task Given by instructor Krish Naik
* Solve in ipynb:
1. Import the dataset.
https://archive.ics.uci.edu/ml/datasets/Algerian+Forest+Fires+Dataset
2. Do proper EDA(analysis) of your dataset and create a report
3. Then perform necessary preprocessing steps


## 2. DATA COLLECTION AND UNDERSTANDING

* I used a dataset on **Algerian Forest Fires** from UCI. The dataset contains a culmination of forest fire observations and data in two regions of Algeria: the Bejaia region and the Sidi Bel-Abbes region. The timeline of this dataset is from **June 2012 to September 2012**. In this project, we focused on whether certain weather features could predict forest fires in these regions using few Classification algorithms.


* As per the Task given i choose regression problem to predict fire weather index

## 3. DATA EXPLORATION

* In this step, we will apply Exploratory Data Analysis (EDA) to extract insights from the data set to know which features have contributed more in predicting Forest fire by performing Data Analysis using Pandas and Data visualization using Matplotlib & Seaborn. It is always a good practice to understand the data first and try to gather as many insights from it.


Below are tasks to be performed in EDA:

**1. Importing Libraries**


**2. Data Cleaning for EDA Report**

**3. Exploratory Data Analysis (EDA) on all Features**


### 3.1 Import Libraries

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
from warnings import filterwarnings
filterwarnings('ignore')

In [3]:
# Create Dataframe and Read the dataset using Pandas
dataset = pd.read_csv('Algerian_forest_fires_dataset_UPDATE.csv', header=1)
dataset.head()

day month  year Temperature  RH  Ws Rain   FFMC  DMC    DC  ISI  BUI  FWI  \
0  01    06  2012          29  57  18     0  65.7  3.4   7.6  1.3  3.4  0.5   
1  02    06  2012          29  61  13   1.3  64.4  4.1   7.6    1  3.9  0.4   
2  03    06  2012          26  82  22  13.1  47.1  2.5   7.1  0.3  2.7  0.1   
3  04    06  2012          25  89  13   2.5  28.6  1.3   6.9    0  1.7    0   
4  05    06  2012          27  77  16     0  64.8    3  14.2  1.2  3.9  0.5   

     Classes    
0  not fire     
1  not fire     
2  not fire     
3  not fire     
4  not fire

In [4]:
# Convert Dataframe into Dictionary as MongoDB stores data in records/documents
data = dataset.to_dict(orient = 'records')

In [5]:
# Insert records in the dataset into MongoDB collection "hotel_records"
db.fire_records.insert_many(data)
print("All the Data has been Exported to MongoDB Successfully")

NameError: name 'db' is not defined

In [6]:
#Convert list into Dataframe
df = pd.DataFrame(list_cursor)
df.drop('_id', axis=1, inplace=True)
df

NameError: name 'list_cursor' is not defined

In [7]:
[features for features in df.columns if df[features].isnull().sum()>1]

NameError: name 'df' is not defined

In [8]:
df.info()

NameError: name 'df' is not defined

## 3.4 Data Cleaning

In [9]:
# Columns which has null values
df[df.isnull().any(axis=1)]

NameError: name 'df' is not defined

In [10]:
df.loc[:122,'Region']=1
df.loc[122:,'Region']=2
df[['Region']] = df[['Region']].astype(int)

NameError: name 'df' is not defined

In [11]:
df.isnull().sum()

NameError: name 'df' is not defined

In [12]:
# Remove null or na values rows
df =df.dropna().reset_index(drop=True)
df.shape

NameError: name 'df' is not defined

In [13]:
# Column which has string
df.iloc[[122]]

NameError: name 'df' is not defined

In [14]:
#remove 122th column
df= df.drop(122).reset_index(drop=True)

NameError: name 'df' is not defined

In [15]:
# List out column names to check
df.columns

NameError: name 'df' is not defined

Looks like the values were misspaced so we can use **str.strip()** method

In [16]:
# Spaces were fixed in the column names
df.columns = df.columns.str.strip()
df.columns

NameError: name 'df' is not defined

In [17]:
**Changing the required columns as integer data type**

SyntaxError: invalid syntax (4034761220.py, line 1)

In [18]:
df[['month', 'day', 'year', 'Temperature','RH', 'Ws']] = df[['month', 'day', 'year', 'Temperature','RH', 'Ws']].astype(int)

NameError: name 'df' is not defined

**Changing the other columns to Float data type**

In [19]:
objects = [features for features in df.columns if df[features].dtypes=='O']
for i in objects:
    if i != 'Classes':
        df[i] = df[i].astype(float)

NameError: name 'df' is not defined

In [20]:
# Final datatypes check
df.info()

NameError: name 'df' is not defined

In [21]:
df.describe().T

NameError: name 'df' is not defined

**Data Cleaning in target variable**

In [22]:
# Check Unique values of target variable
df.Classes.value_counts()

NameError: name 'df' is not defined

Looks like the values were misspaced so we can use **str.strip()** method

In [23]:
# Strip mispaced values
df.Classes = df.Classes.str.strip()

NameError: name 'df' is not defined

In [24]:
df.Classes.value_counts()

NameError: name 'df' is not defined

#### Region 1 (Bejaia Region)

In [25]:
df[:122]

NameError: name 'df' is not defined

In [27]:
#### Region 2 (Sidi-Bel Abbes Region)
df[122:]

NameError: name 'df' is not defined

In [28]:
df.to_csv('Algerian_forest_fires_dataset_CLEANED.csv', index=False)

NameError: name 'df' is not defined

### 3.5 Exploratory Data Analysis (EDA)

In [29]:
#Droping Year features
df1 = df.drop(['day','month','year'], axis=1)

NameError: name 'df' is not defined

In [30]:
# Encoding Not fire as 0 and Fire as 1
df1['Classes']= np.where(df1['Classes']== 'not fire',0,1)

NameError: name 'df1' is not defined

In [31]:
# Check counts
df1.Classes.value_counts()

NameError: name 'df1' is not defined

**Visualization of Distribution**

In [32]:
# PLot density plot for all features
plt.style.use('seaborn')
df1.hist(bins=50, figsize=(20,15), ec = 'b')
plt.show()

NameError: name 'df1' is not defined

In [33]:
# Percentage for PieChart
percentage = df.Classes.value_counts(normalize=True)*100
percentage

NameError: name 'df' is not defined

**Charts**

In [34]:
#plotting PieChart
classeslabels = ["FIRE", "NOT FIRE"]
plt.figure(figsize =(12, 7))
plt.pie(percentage,labels = classeslabels,autopct='%1.1f%%')
plt.title ("Pie Chart of Classes", fontsize = 15)
plt.show()

NameError: name 'percentage' is not defined

<Figure size 1200x700 with 0 Axes>

**Checking Multicollinearity**

In [35]:
# Correlation chart
k = len(df1.columns)
cols = corr.nlargest(k, 'Classes')['Classes'].index
cm = np.corrcoef(df1[cols].values.T)
sns.set(font_scale=1)
f, ax = plt.subplots(figsize=(20, 13))
hm = sns.heatmap(cm, annot=True, square=True, fmt='.2f', annot_kws={'size': 15}, yticklabels=cols.values, xticklabels=cols.values)
plt.show()

NameError: name 'df1' is not defined

#### Canadian Forest Fire Weather Index (FWI) System
This column can be considered for Regression problem

In [36]:
#Boxplot
ax = sns.boxplot(df['FWI'], color= 'red')

NameError: name 'df' is not defined

### **Monthwise Fire Analysis**

In [37]:
dftemp= df.loc[df['Region']== 1]
plt.subplots(figsize=(13,6))
sns.set_style('whitegrid')
sns.countplot(x='month',hue='Classes',data= df,ec = 'black', palette= 'Set2')
plt.title('Fire Analysis Month wise for Bejaia Region', fontsize=18, weight='bold')
plt.ylabel('Count', weight = 'bold')
plt.xlabel('Months', weight= 'bold')
plt.legend(loc='upper right')
plt.xticks(np.arange(4), ['June','July', 'August', 'September',])
plt.grid(alpha = 0.5,axis = 'y')
plt.show()

NameError: name 'df' is not defined

In [38]:
dftemp= df.loc[df['Region']== 2]
plt.subplots(figsize=(13,6))
sns.set_style('whitegrid')
sns.countplot(x='month',hue='Classes',data= df,ec = 'black', palette= 'Set2')
plt.title('Fire Analysis Month wise for Sidi-Bel Abbes Region', fontsize=18, weight='bold')
plt.ylabel('Count', weight = 'bold')
plt.xlabel('Months', weight= 'bold')
plt.legend(loc='upper right')
plt.xticks(np.arange(4), ['June','July', 'August', 'September',])
plt.grid(alpha = 0.5,axis = 'y')
plt.show()

NameError: name 'df' is not defined

* Its observed that August and September had the most number of forest fires for both regions. And from the above plot of months, we can understand few things

* Most of the fires happened in August and very high Fires happened in only 3 months - **June, July and August.**

* Less Fires was on September

### 3.5 Exploratory Data Analysis (EDA) on all Features

In [39]:
df.columns

NameError: name 'df' is not defined

In [40]:
def barchart(feature,xlabel):
    plt.figure(figsize=[14,8])
    by_feature =  df1.groupby([feature], as_index=False)['Classes'].sum()
    ax = sns.barplot(x=feature, y="Classes", data=by_feature[[feature,'Classes']], estimator=sum)
    ax.set(xlabel=xlabel, ylabel='Fire Count')

In [41]:
barchart('Temperature','Temperature Max in Celsius degrees')

NameError: name 'df1' is not defined

<Figure size 1400x800 with 0 Axes>

In [42]:
barchart('Rain', 'Rain in mm')

NameError: name 'df1' is not defined

<Figure size 1400x800 with 0 Axes>

#### Fine Fuel Moisture Code (FFMC) index

In [43]:
barchart('Ws', 'Wind Speed in km/hr')

NameError: name 'df1' is not defined

<Figure size 1400x800 with 0 Axes>

In [44]:
barplots('RH','Relative Humidity in %')

NameError: name 'barplots' is not defined

In [45]:
dftemp = df1.drop(['Classes', 'Region'], axis=1)
fig = plt.figure(figsize =(12, 6))
ax = dftemp.boxplot()
ax.set_title("Boxplot of Given Dataset")
plt.show()

NameError: name 'df1' is not defined

In [46]:
dftemp = dftemp = df1.drop(['Region','Temperature','Rain','Ws','RH'], axis=1)
for feature in dftemp:
    sns.histplot(data = dftemp,x=feature, hue = 'Classes')
    plt.legend(labels=['Fire','Not Fire'])
    plt.title(feature)
    plt.show()

NameError: name 'df1' is not defined

### REPORT
#### Weather System Report
* **Temperature** Highest Fire counts happened between **30-37 degree Celsius**
* **Rain** Highest Fire counts happened when there was no rain to very less rain ie. **0.0 to 0.3.**
* **Wind Speed** highest Fire count happened when the wind speed were between **13 to 19 Km/hr.**
* **Relative Humidity** highest fire count happened when the RH is between **50 to 80%.**


#### FWI System Components Report
FWI (Canadian Forest Fire Weather Index)
* **Fine Fuel Moisture Code (FFMC) index** which ranges between *28.6 to 92.5*,
    here **above 75** has higher chance of Forest fires.
    
* **Duff Moisture Code (DMC) index** which ranges between *1.1 to 65.9*,
    here 1.1-10 has lower chance of Forest fires whereas above **10-30 DMC** has very high evidence of Forest fires in past.
    
* **Drought Code (DC) index** which ranges between *7 to 220.4*,
    here 0-25 is safe and has lower chance of Forest fires whereas range **above 25 DC** has higher chance of forest fires.
    
* **Initial Spread Index (ISI) index** which ranges between *0 to 18*,
    here 0-3 has lower Forest fires and **above 3 ISI** has higher chance of Forest fires.
    
* **Buildup Index (BUI) index** which ranges between *1.1 to 68*,
    here 1.1 to 10 has lower Forest fire chance and **above 10 BUI has** higher chance of forest fires.
    
* **Fire Weather Index (FWI) Index** which ranges between *1 to 31.1*,
    here 0-3 has lower chance of Forest fires and **3-25 FWI** has higher chance of forest fires.